### PR TITLE
Identity resolution for loading detached entities

### PR DESCRIPTION
**Pull request by ajcvickers**
Part of #10042

Avoids duplicates when loading collections that already contain some entities. Entities queried using NoTrackingWithIdentityResolution get this behavior when lazy-loading.

### CODE REVIEW

File: src/EFCore/ChangeTracking/CollectionEntry.cs [hunk 1]
```diff
@@ -254,7 +300,41 @@ public override Task LoadAsync(CancellationToken cancellationToken = default)
 
         return IsLoaded
             ? Task.CompletedTask
-            : TargetLoader.LoadAsync(InternalEntry, cancellationToken);
+            : TargetLoader.LoadAsync(InternalEntry, forceIdentityResolution: false, cancellationToken);
+    }
+
+    /// <summary>
+    ///     Loads entities referenced by this navigation property, unless <see cref="NavigationEntry.IsLoaded" />
+    ///     is already set to true.
```
Comment (AndriySvyryd) [inline src/EFCore/ChangeTracking/CollectionEntry.cs:308]:
```suggestion
    ///     is already set to <see langword="true" />.
```
And elsewhere

File: src/EFCore/Internal/EntityFinder.cs [hunk 2]
```diff
@@ -504,31 +529,119 @@ await foreach (var loaded in queryable.AsAsyncEnumerable().WithCancellation(canc
...
+                    && forceIdentityResolution
+                    && TryGetTracked(out var existing))
+                {
+                    value = existing;
+                }
+
                 entry.SetProperty(navigation, value, isMaterialization: true, setModified: false);
             }
+
+            bool TryGetTracked(out object? tracked)
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/EntityFinder.cs:569]: Make this and the nested one static for perf

File: src/EFCore/Internal/EntityFinder.cs [hunk 3]
```diff
@@ -417,26 +418,37 @@ private bool TryFindByKey(IReadOnlyList<IProperty> properties, IReadOnlyList<obj
...
+                        Fixup(stateManager, entry.Entity, navigation, inverse, forceIdentityResolution, queryable.FirstOrDefault());
                     }
                 }
-                else
+                finally
                 {
-                    Fixup(entry, navigation, inverse, queryable.FirstOrDefault());
+                    if (stateManager != _stateManager)
+                    {
+                        stateManager.ResetState();
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/EntityFinder.cs:450]: Why is this needed if this state manager is discarded?

File: src/EFCore/Internal/EntityFinder.cs [hunk 4]
```diff
@@ -504,31 +529,119 @@ await foreach (var loaded in queryable.AsAsyncEnumerable().WithCancellation(canc
...
+                        propertyValues[i] = propertyValue;
+                    }
+
+                    return propertyValues;
+                }
+            }
         }
     }
 
+    private IStateManager GetOrCreateStateManager(
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/EntityFinder.cs:610]: Refactor this into two methods for readability

File: src/EFCore/Internal/ManyToManyLoader.cs [hunk 5]
```diff
@@ -95,22 +119,84 @@ public virtual async Task LoadAsync(InternalEntityEntry entry, CancellationToken
...
+        {
+            entry.AddToCollection(_skipNavigation, existing!, forMaterialization: true);
+        }
+        else
+        {
+            entry.AddToCollection(_skipNavigation, loaded, forMaterialization: true);
+            relatedEntry.AddToCollection(_inverseSkipNavigation, entity, forMaterialization: true);
+        }
+
+        bool TryGetTracked(out object? tracked)
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/ManyToManyLoader.cs:138]: Consider moving these helpers to StateManager to avoid duplication

File: test/EFCore.Specification.Tests/ManyToManyLoadTestBase.cs [hunk 6]
```diff
@@ -350,6 +350,165 @@ public virtual async Task Load_collection_using_Query_already_loaded(EntityState
...
+    [InlineData(EntityState.Unchanged, true, false)]
+    [InlineData(EntityState.Added, true, true)]
+    [InlineData(EntityState.Added, true, false)]
+    [InlineData(EntityState.Modified, true, true)]
+    [InlineData(EntityState.Modified, true, false)]
+    [InlineData(EntityState.Deleted, true, true)]
+    [InlineData(EntityState.Deleted, true, false)]
+    [InlineData(EntityState.Detached, true, true)]
+    [InlineData(EntityState.Detached, true, false)]
+    public virtual async Task Load_collection_partially_loaded(EntityState state, bool forceIdentityResolution, bool async)
```
Comment (AndriySvyryd) [inline test/EFCore.Specification.Tests/ManyToManyLoadTestBase.cs:374]: Also add a test that uses a property bag for the join entity type

**Review by AndriySvyryd — APPROVED** [review]

File: src/EFCore/Internal/EntityFinder.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore/Internal/EntityFinder.cs:450]: It unsubscribes notification entities.

File: src/EFCore/Internal/EntityFinder.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore/Internal/EntityFinder.cs:569]: Does this means that we should not, in general, use local functions that reference variables from the enclosing function? This is, to me, one of the main advantages of local functions.

File: src/EFCore/Internal/EntityFinder.cs (hunk 4 above)
Comment (ajcvickers) [reply src/EFCore/Internal/EntityFinder.cs:610]: What do the two methods do?

File: src/EFCore/Internal/EntityFinder.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Internal/EntityFinder.cs]: No, only on hot paths

File: src/EFCore/Internal/EntityFinder.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore/Internal/EntityFinder.cs:586]: GetOrCreateStateManager and StartTracking
